Display Metadata

MVarshini opened the pull request:
PBENCH-115

Display Metadata

On row expansion display the metadata

webbnh commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
+      const isParentChecked = findItemCheked(aggregateKey, checkedItems);
+      const newKey = `${aggregateKey}${CONSTANTS.KEYS_JOIN_BY}${item}`;
+      const isChecked = isParentChecked || findItemCheked(newKey, checkedItems);
+      if (isChecked && !checkedItems.includes(newKey)) {
+        dispatch(updateCheckedItems(newKey));
+      }
```
I'm confused and concerned by how this might be working.  We're updating the Redux store's instance of the checked items list, but we're not updating a local copy of it.  And, the update is dispatched to a task, so it doesn't happen right away.  So, every time we look at `checkedItems` we're looking at the state from the beginning of the parse.

It's probably OK that we're not keeping it up to date, and it's probably fine to defer the updates, but it seems inefficient to queue a pile of updates when, instead, we could just create a local copy of the state, update it as we go, and then update the Redux store _once_ at the end of the parse.

(And, if we pass in a reference to the list of checked items, then updates here would be visible in the caller; and, we wouldn't need to pass in `dispatch()` because the caller can make the call.)

MVarshini replied:
I'm not sure where to maintain the checkItems list during parsing

webbnh commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
+
+  for (const item of optionRoles) {
+    const dataObj = { title: item, options: [] };
+    for (const key in keySummary[item]) {
+      if (!isNonEssentialKey(key)) {
+        const aggregateKey = `${item}${CONSTANTS.KEYS_JOIN_BY}${key}`;
+        const isChecked = findItemCheked(aggregateKey, checkedItems);
+        const obj = constructTreeObj(aggregateKey, isChecked);
+        if (!keySummary[item][key]) {
+          dataObj.options.push(obj);
+        } /* has children */ else {
+          obj["children"] = constructChildTreeObj(
+            aggregateKey,
+            keySummary[item][key],
+            checkedItems,
+            dispatch
+          );
+          dataObj.options.push(obj);
+        }
+      }
+    }
```
How is this code different from `constructChildTreeObj()`?

I think the differences are around updating the checked items list (which I'm hoping you will defer until the end of the parse) and the construction of the "aggregate key" (which I think we can handle with a local conditional); if we can address those differences, then I don't think we need to duplicate this code.

webbnh commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
+  return childObj;
+};
+
+const constructTreeObj = (aggregateKey, isChecked) => {
+  const obj = {};
+  const keyArr = aggregateKey.split(CONSTANTS.KEYS_JOIN_BY);
+  obj["name"] = keyArr[keyArr.length - 1];
+  obj["key"] = aggregateKey;
+  obj["id"] = aggregateKey;
+  obj["checkProps"] = {
+    "aria-label": `${aggregateKey}-check`,
+    checked: isChecked,
+  };
+  return obj;
+};
```
Alternatively,
```javascript
const constructTreeObj = (aggregateKey, isChecked) => (
  {
    "name": aggregateKey.split(CONSTANTS.KEYS_JOIN_BY).pop(),
    "key": aggregateKey,
    "id": aggregateKey,
    "checkProps": {
      "aria-label": `${aggregateKey}-check`,
      checked: isChecked,
    }
  }
);
```

webbnh commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
+  CONSTANTS.KEY_TOOLS_REGEX,
+  CONSTANTS.KEY_ITERATIONS_REGEX,
+];
+const isNonEssentialKey = (string) => {
+  const len = nonEssentialKeys.length;
+  for (let i = 0; i < len; i++) {
+    if (string.match(nonEssentialKeys[i])) {
+      return true;
+    }
+  }
+  return false;
+};
```
Alternately,
```javascript
const isNonEssentialKey = (string) => !!nonEssentialKeys.find((e) => string.match(e))
```

MVarshini replied:
Updated

webbnh reviewed (commented):
I only got about halfway through, but here's what I've got so far.  Most of these are coding suggestions.  In a couple of places, it seems like we're using the Redux store where we don't need to or where we could update it once instead of repeatedly.

More later.

webbnh commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -0,0 +1,123 @@
+    const checked = evt.target.checked;
+
+    const treeData = getState().overview.treeData;
+    let childKeys = [];
```
`childKeys` doesn't seem to be referenced outside the block at line 75 where it is assigned; I recommend declaring it in that block and using `const` instead of `let`.

MVarshini replied:
Done

webbnh commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -0,0 +1,123 @@
+    const isFirstChild = keys.length === 2;
+    const isParent = "children" in treeViewItem;
+    if (isParent) {
+      const childNodes = treeViewItem.children;
+      childKeys = getCheckedItemsKey(childNodes);
+      const modifiedChildNodes = setChildNodes(childNodes, checked);
+      const idx = options.findIndex((item) => keys.includes(item.name));
+
+      if (idx > -1) {
+        const childNodes = options[idx]["children"];
```
Line 82 defines another variable with the same name as the variable defined at line 76.  This seems like a bad idea.

MVarshini replied:
Modified the function

webbnh commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -0,0 +1,123 @@
+      type: TYPES.SET_TREE_DATA,
+      payload: treeData,
+    });
+    dispatch({
+      type: TYPES.SET_METADATA_CHECKED_KEYS,
+      payload: checkedItems.flat(),
+    });
```
By default, `flat()` flattens only one level...is that what we want here, or should we be specifying `Infinity`?

MVarshini replied:
Now,`flat()` is not required. As new element is added to `checkedItems`  with `spread(...)` operator instead of being pushed

webbnh commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -0,0 +1,123 @@
+        const childNodes = options[idx]["children"];
+        const cIdx = getIndexofKey(childNodes, treeViewItem.key);
+        if (cIdx > -1) {
+          childNodes[cIdx]["children"] = [...modifiedChildNodes];
```
I believe that line 85 is modifying the Redux store, which I think we're supposed to avoid.  (Ditto lines 86, 93, 105.)  So, presumably we should be building and modifying a copy of `treeData` which we then send to the Redux store at line 117.

webbnh reviewed (changes requested):
I've got a few more coding suggestions and similar comments (which you are free to decline...); however, I found a few items which pose more concern.
- Foremost, in `mapTree()` in `dashboard/src/actions/metadataTreeActions.js`, I think the code is overwriting the Redux store, which we are supposed to avoid.
- In `areAllDescendantsChecked()` and `areSomeDescendantsChecked()` there seems to be a difference in behavior depending on whether the node has children; if the code is correct as is, it would be good to have a comment explaining why it does that.
- And there are some smaller items which deserve attention (like using `map()` vs. `forEach()`, `find()` vs. `findIndex()`, a nested variable declaration with the same name, and a function which sometimes returns a new object and sometimes returns an old one).

And, I've got a couple of pointed questions.  🙂

MVarshini commented on dashboard/src/modules/components/OverviewComponent/NewRunsComponent.jsx:
```diff
@@ -166,29 +178,45 @@ const NewRunsComponent = () => {
-                      moreActionItems={moreActionItems}
-                      setRunExpanded={setRunExpanded}
-                      isRunExpanded={isRunExpanded}
-                      toggleEdit={toggleEdit}
-                      onSelectRuns={onSelectRuns}
-                      isRowSelected={isRowSelected}
-                      columnNames={columnNames}
-                      makeFavorites={makeFavorites}
-                      saveRowData={saveRowData}
-                      rowActions={rowActions}
-                      item={item}
-                      textInputEdit={(val) =>
-                        updateTblValue(val, NAME_KEY, item.resource_id)
-                      }
-                    />
-                  </Tr>
+                      className={item[IS_ITEM_SEEN] ? "seen-row" : "unseen-row"}
+                    >
+                      <NewRunsRow
+                        key={item.resource_id}
```
`key` is used by React to identify the DOM that is being updated so it can be re-rendered.

MVarshini commented:
I have addressed most of the comments and the requested changes

* The looseness in the regex for the keys to exclude
* Generated comments for critical functions using JDoc
* In few places moved from `findIndex` to `find` 
* Declaration of unnecessary variables  
* For `areSomeDescendantsChecked()` and `areAllDescendantsChecked()`, the `checkedItems` array contains the parent of the checked children irrespective the all the children being checked or not

dbutenhof commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -525,16 +522,15 @@ const updateCheckedItems = (keyToInclude) => (dispatch, getState) => {
-
-const findItemCheked = (key, checkedItems) =>
-  checkedItems.includes(key) ? true : false;
+const findItemsCheked = (key, checkedItems) => checkedItems.includes(key);
```
`findItemsChecked`?

dbutenhof commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -428,42 +426,43 @@ export const getKeySummary = async (dispatch, getState) => {
     const dataObj = { title: item, options: [] };
     for (const key in keySummary[item]) {
-      if (!isNonEssentialKey(key)) {
+      if (!isServerInternal(`${item}${CONSTANTS.KEYS_JOIN_BY}${key}`)) {
```
If `KEYS_JOIN_BY` is `*`, then we try to check `server*index-map`, which doesn't match your intended pattern ... except that `.` is the regex wildcard so you're actually allowing more patterns we don't want to match. More on that below. 😆

dbutenhof reviewed (changes requested):
We still have problems with the "excluded pattern" matches. On the one hand, undesired matches of your loose patterns may be unlikely; but since we don't have a "raw unfiltered" mode we risk blocking real user metadata keys someone might need.

Another option might be to simply allow everything that starts with `/^global\./` or `/^user\./`; but just tightening up your patterns would be simpler.

webbnh reviewed (commented):
My comments below are just code-quality suggestions.  However, there are a number of comments that I made on my earlier passes that you haven't responded to in the "conversations" and you haven't addressed in the code (so, I don't know whether you actually saw them).

The ones that I would most like to hear your thoughts on are these:
- The Redux store updates [during the parse](https://github.com/distributed-system-analysis/pbench/pull/3464#discussion_r1231533959)
- [This code](https://github.com/distributed-system-analysis/pbench/pull/3464#discussion_r1231546181) which nearly duplicates `constructChildTreeObj()`
- The [behaviors of `areAllDescendantsChecked()`](https://github.com/distributed-system-analysis/pbench/pull/3464#discussion_r1232502864) and `areSomeDescendantsChecked()`
- The [lack of recursion](https://github.com/distributed-system-analysis/pbench/pull/3464#discussion_r1232510895) in `setChildNodes()`
- The [use of `map()`](https://github.com/distributed-system-analysis/pbench/pull/3464#discussion_r1232516739) instead of `forEach()`
- The [use of `flat()`](https://github.com/distributed-system-analysis/pbench/pull/3464#discussion_r1232589720) with no parameter.

And, I'd like to see [this thunk](https://github.com/distributed-system-analysis/pbench/pull/3464#discussion_r1231526843) converted to an action.

webbnh reviewed (commented):
I don't see any new changes here since my last review on 27 June.

MVarshini commented on dashboard/src/assets/constants/overviewConstants.js:
```diff
@@ -18,6 +18,11 @@ export const IS_DIRTY_SERVER_DELETE = "isDirtyDate";
+export const KEY_INDEX_REGEX = /index-map/;
+export const KEY_ITERATIONS_REGEX = /iterations/;
+export const KEY_OPERATIONS_REGEX = /operations/;
+export const KEY_TOOLS_REGEX = /tools/;
```
Updated the regex patterns

MVarshini commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
+  }
+};
+
+export const parseKeySummaryforTree = () => (dispatch, getState) => {
```
Added the JDoc comments

MVarshini commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
+  CONSTANTS.KEY_TOOLS_REGEX,
+  CONSTANTS.KEY_ITERATIONS_REGEX,
+];
+const isNonEssentialKey = (string) => {
```
Updated the name to be  `internal server`

MVarshini commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
+    type: TYPES.SET_METADATA_CHECKED_KEYS,
+    payload: checkedItems,
+  });
+};
```
Updated

MVarshini commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
+  const optionRoles = Object.keys(keySummary);
+  const checkedItems = getState().overview.checkedItems;
+
+  for (const item of optionRoles) {
```
Got it

MVarshini commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
+
+  for (const item of optionRoles) {
+    const dataObj = { title: item, options: [] };
+    for (const key in keySummary[item]) {
```
Gotcha

MVarshini commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
+      const isChecked = isParentChecked || findItemCheked(newKey, checkedItems);
+      if (isChecked && !checkedItems.includes(newKey)) {
+        dispatch(updateCheckedItems(newKey));
+      }
```
Updated

MVarshini commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -0,0 +1,123 @@
+const areAllDescendantsChecked = (dataItem, checkedItems) =>
+  dataItem.children
+    ? dataItem.children.every((child) => child.checkProps.checked === true)
+    : isChecked(dataItem, checkedItems);
```
dataItem checked value depends on its children.So, it need not be checked

webbnh replied:
So...the only values which matter are the leaf nodes?...all the intermediate nodes are basically ignored?

MVarshini commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -0,0 +1,123 @@
+    const keys = treeViewItem.key.split(KEYS_JOIN_BY);
+    const isFirstChild = keys.length === 2;
+    const isParent = "children" in treeViewItem;
+    if (isParent) {
```
Updated

MVarshini commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -0,0 +1,123 @@
+  childNodes.map((item) => {
+    item.checkProps.checked = isChecked;
+    return item;
+  });
```
Right! Modified the code

MVarshini commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -0,0 +1,123 @@
+  childNodes.map((item) => {
+    item.checkProps.checked = isChecked;
+    return item;
+  });
```
Updated with `for each`

MVarshini commented on dashboard/src/assets/constants/overviewConstants.js:
```diff
@@ -19,10 +19,11 @@ export const IS_EDIT = "isEdit";
+export const KEY_ITERATIONS_REGEX = /dataset.metalog.iterations/;
+export const KEY_OPERATIONS_REGEX = /dataset.operations/;
+export const KEY_TOOLS_REGEX = /dataset.metalog.tools/;
+export const KEY_TARBALL_PATH_REGEX = /server.tarball-path/;
```
Updated the regex pattern

MVarshini commented on dashboard/src/modules/components/OverviewComponent/common-component.jsx:
```diff
@@ -387,23 +389,25 @@ export const MetaDataModal = () => {
+                    <span className="keyClass">
+                      {levels[levels.length - 1]}
+                    </span>{" "}
+                    : {showValue}
```
`<span>` is used to add class to the key so css can be applied. `{" "}` for specifying the spaces

webbnh commented on dashboard/src/assets/constants/overviewConstants.js:
```diff
@@ -18,6 +18,12 @@ export const IS_DIRTY_SERVER_DELETE = "isDirtyDate";
 export const IS_ITEM_FAVORITED = "isItemFavorited";
 export const IS_ITEM_SEEN = "isItemSeen";
+export const KEYS_JOIN_BY = "*";
+export const KEY_INDEX_REGEX = /^server[.*]index-map$/;
+export const KEY_ITERATIONS_REGEX = /^dataset[.*]metalog[.*]iterations/;
+export const KEY_OPERATIONS_REGEX = /^dataset[.*]operations$/;
+export const KEY_TOOLS_REGEX = /^dataset[.*]metalog[.*]tools/;
+export const KEY_TARBALL_PATH_REGEX = /^server[.*]tarball-path$/;
```
> I don't like the wildcarded separator pattern... _and_ the `[]` grouping is meaningless here since there's a single character inside.

@dbutenhof, I think you might be confused.  I think the `[]` means "match any (one) of these (literal) characters", so we're matching on the keys separated by either a _period_ **or** an _asterisk_ -- there's no wildcard here (other than the alternation), and there are _two_ characters in the set.

(Does that change any of your other comments, here?)

dbutenhof reviewed (changes requested):
I pulled the branch to experiment a bit, and there are some problems.

![image](https://github.com/distributed-system-analysis/pbench/assets/61761529/6ad45dee-6c3c-4024-ae77-1067058931d2)

Note that there's no value for "Archiveonly" -- it's a *boolean* value, which may have something to do with that.

Also, both of the two values are shown with just the leaf key name, capitalized, "Archiveonly" (`server.archiveonly`) and `Benchmark` (`server.benchmark`). I think I'd rather see the full key name. For example, I created `global.dave` and `global.foobar.dave`: assuming that selection actually worked, both would show as just "Dave" and there'd be no way to tell them apart in the output!

I also added some `global` metadata keys; and while I can select them in the modal tree,

![image](https://github.com/distributed-system-analysis/pbench/assets/61761529/ec6728eb-6d90-4f51-9023-4fd001157841)

They don't appear in the output

![image](https://github.com/distributed-system-analysis/pbench/assets/61761529/ec0b9ed2-0dee-46b0-8f87-cb4b1d574179)

dbutenhof commented:
OK, that's a lot. To approve,

1. We need to be able to see metadata across all the namespaces, including global.
2. We need to be able to see metadata that's not a simple string, even if it's just a simple `string(value)` conversion.

Other stuff I'd prefer to work differently, after experimenting, but I'm happy not to gate this PR on the rest. For example,

I'm not entirely sure I like the pre-selection of `dataset.metalog.run` and `dataset.access`, but that's a detail. Some sort of hierarchical representation of the metadata you show, instead of just the "leaf" of the key path, would be good, whether it's an indented list like

```
global
    foobar
        dave: "foobar"
    dave: "foobar"
```

or just a linear list of the full name, like `global.foobar.dave`: although that would make your `dataset.metalog.run` look a lot messier than

```
dataset
    access: public
    metalog:
        run:
            controller : ctlrA
            end_run : 2020-02-28T19:16:58.350741523
            prefix : linpack-dev2
            raw_size : 86906351
            start_run : 2020-02-28T19:10:55.645002669
            user : janedo@example.com
```

webbnh reviewed (changes requested):
I think we need to discuss disabling the lint rule before this PR merges.

Other than that, I have some questions, a nit, and some coding suggestions for you to consider.  I've closed out my previous conversations which have been resolved, and posted follow-ups to the others.

webbnh commented:
> Some sort of hierarchical representation of the metadata you show, instead of just the "leaf" of the key path, would be good

On the output side, while we certainly need the whole path represented (so that we can disambiguate two keys with the same name but different paths), I'm not sure that we really want a hierarchical display -- if we have very many results in the query, we're going to want some way to show them which is reasonably compact, I think.  (I'm thinking a columnar display, which will serve as a disincentive to trying to view too many values at a time, while supporting large numbers of results and enabling sorting and other filtering based on the values.)

dbutenhof commented:
> On the output side, while we certainly need the whole path represented (so that we can disambiguate two keys with the same name but different paths), I'm not sure that we really want a hierarchical display -- if we have very many results in the query, we're going to want some way to show them which is reasonably compact, I think. (I'm thinking a columnar display, which will serve as a disincentive to trying to view too many values at a time, while supporting large numbers of results and enabling sorting and other filtering based on the values.)

Basically my concern is

```
dataset
    access: public
    metalog:
        run:
            controller: ctlrA
            end_run: 2020-02-28T19:16:58.350741523
            prefix: linpack-dev2
            raw_size: 86906351
            start_run: 2020-02-28T19:10:55.645002669
            user: janedo@example.com
```

vs the current

```
Access: public
Controller: ctlrA
End_run: 2020-02-28T19:16:58.350741523
Prefix: linpack-dev2
Raw_size: 86906351
Start_run: 2020-02-28T19:10:55.645002669
User: janedo@example.com
```

vs something like

```
dataset.access: public
dataset.metalog.run.controller: ctlrA
dataset.metalog.run.end_run: 2020-02-28T19:16:58.350741523
dataset.metalog.run.prefix: linpack-dev2
dataset.metalog.run.raw_size: 86906351
dataset.metalog.run.start_run: 2020-02-28T19:10:55.645002669
dataset.metalog.run.user: janedo@example.com
```

In comparison, the indented form seems a lot easier to read. If I'm correctly inferring what you mean, yeah, column alignment might help, even if I'm not thrilled about the whitespace gaps we can get in a "Key" column with short paths contrasted with relatively long paths.

```
Key                                     Value
--------------------------------------- -----------------------------
dataset.access                          public
dataset.metalog.run.controller          ctlrA
dataset.metalog.run.end_run             2020-02-28T19:16:58.350741523
dataset.metalog.run.prefix              linpack-dev2
dataset.metalog.run.raw_size            86906351
dataset.metalog.run.start_run           2020-02-28T19:10:55.645002669
dataset.metalog.run.user                janedo@example.com
global.project.run.host.user.breakfast  eggs
```

And I really don't want a "disincentive" to view metadata: I want to make it easy and natural and, ideally, somewhat useful; possibly with a UI like being able to select (including multi-select) values for filtering directly from the display. (Which also implies a selection UI to edit the match term for wildcarding and type: the Opensearch +/- term filter chips might be a good start, although it'd be nice to avoid relying on manually typing `AND NOT message:"substr"` where we don't want a full literal match on the field.)

(And, also, the current metadata display is active only in the Overview list, whereas I think we need to extend it to the browsing page ... but the more limited scope of this PR is fine to start.)

webbnh commented:
Actually, I think the answer is very context-sensitive.

If the user has one result (or a small enough number of results that the display doesn't require a scroll bar), then I think we have a wide latitude for display options, and I think that displaying the metadata in some sort of hierarchical fashion (as you've proposed) makes excellent sense.  (I might be inclined toward an accordion arrangement like the selectors, particularly if there are lots of keys, but that's neither here nor there.)

My concern runs toward the case where the user is looking at 10s or 100s of results.  In that case, I'm thinking that the display should be more of a table, with each dataset occupying a row and each selected metadata key heading a column with the key's value in the corresponding cell.  This is likely to produce a very wide display (e.g., if the user casually requests all of the `dataset` or `dataset.metalog` keys), and so, to make it useful, the onus will be on the user to be quite specific about what s/he wants to see.  But, once that selection is made, then the format will enable things like clicking on column headings to sort the results.  And, we could, as you say, borrow from Kibana/Opensearch to allow the user to easily filter on/out specific values.

But, as you say, this should be available in both displays, but not in this PR.  🙂

dbutenhof commented:
> In that case, I'm thinking that the display should be more of a table, with each dataset occupying a row and each selected metadata key heading a column with the key's value in the corresponding cell.

First, the display of metadata is attached to and indented under each dataset. You can open one or all datasets to reveal the selected metadata values. A table such as you describe could easily become as wide and unwieldy as the old dashboard "detail" views, with scrolling in both directions making it almost impossible to figure out what's what.

So effectively we have

```
dataset-name
🔻 dataset-name
  dataset.access: public
  global.dashboard.saved: True
🔻 dataset-name
  dataset.access: private
  global.dashboard.saved: False
dataset-name
dataset-name
```

webbnh commented:
> the display of metadata is attached to and indented under each dataset.

Right, and this is likely to be counterproductive (or even annoying) when the number of datasets in the display is large.

> A table such as you describe could easily become as wide and unwieldy as the old dashboard "detail" views, with scrolling in both directions making it almost impossible to figure out what's what.

Indeed, and that is why the user would need to be selective about which metadata s/he includes in the display.

> So effectively we have
> 
> ```
> dataset-name
> 🔻 dataset-name
> ...
> ```

Right, and so we'll need some sort of policy/controls on which items appear "expanded" initially/by-default...otherwise, a display with more than a few results will be awkward to read until all the initially unwanted keys have been collapsed.

My point is, for the purpose of viewing a single dataset (or perhaps a few), I think what you're proposing here is perfect...but I don't think it scales when the number of datasets being displayed no longer fits on the screen, and I think we're going to want an alternative approach in that case.

MVarshini commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -0,0 +1,103 @@
+    ? dataItem.children.some((child) =>
+        areSomeDescendantsChecked(child, checkedItems)
+      )
+    : isChecked(dataItem, checkedItems);
```
Yes, it should be. Thanks for pointing it out

MVarshini commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -52,50 +57,38 @@ export const mapTree = (item) => {
-      const childIdx = getIndexofKey(childNodes, parentKey);
+      checkedItems = updateChildKeysList(checked, checkedItems, childKeys);
+    } else if ("children" in options[0]) {
+      /* if first child*/ const childNodes = options[0]["children"];
```
Got it

MVarshini commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -52,50 +57,38 @@ export const mapTree = (item) => {
-    } else if (isFirstChild) {
-      const realOption = options.find((item) => item.id === treeViewItem.id);
+      setChildNodes(childNodes, checked);
+      treeViewItem.checkProps.checked = checked;
```
Updated

MVarshini commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -23,6 +27,8 @@ export const getDatasets = () => async (dispatch, getState) => {
     params.append("metadata", CONSTANTS.DATASET_UPLOADED);
     params.append("metadata", CONSTANTS.SERVER_DELETION);
     params.append("metadata", CONSTANTS.USER_FAVORITE);
+    params.append("metadata", "server");
+    params.append("metadata", "dataset");
```
Let's have this "dataset.uploaded"  "server.deletion so I can avoid few more changes.

Now, added "global" and "user" params

dbutenhof commented:
> My point is, for the purpose of viewing a single dataset (or perhaps a few), I think what you're proposing here is perfect...but I don't think it scales when the number of datasets being displayed no longer fits on the screen, and I think we're going to want an alternative approach in that case.

We have a fundamental disagreement here, but it doesn't matter for this PR and may never matter so there's probably no point in belaboring it now.

dbutenhof reviewed (changes requested):
A strong step closer, but I see two problems... the second probably isn't worth attacking here, but I'll point it out anyway.

This is still only showing string values correctly: `server.archiveonly` `global.dashboard.saved`, `global.dashboard.seen`, and `user.dashboard.favorite` are boolean values, and are displayed as blanks. Any key value in the `user` or `global` space could have a non-string value (int, bool, float, list, etc.), so ideally we should be able to make at least a valiant attempt to display it. (Which is why I suggested a simple string-ification.)

![image](https://github.com/distributed-system-analysis/pbench/assets/61761529/c9373576-fd4f-458b-b49b-cffb5f0af22a)

There's also a refresh problem, although I wouldn't hold this PR to fix it. Basically, the dashboard has a mechanism on the Overview page to *set* metadata ... specifically `user.dashboard.favorite`, `global.dashboard.seen`, and `global.dashboard.saved`. *Changing* any of these doesn't cause the component to update the `keysummary` or the cached metadata, though, meaning that once I mark a dataset "seen" I ought to see `global.dashboard.seen` appear in the modal selector, (and then in the list where appropriate), but I have to navigate away from the page and back to get that to happen.

webbnh commented on dashboard/src/actions/metadataTreeActions.js:
```diff
@@ -0,0 +1,104 @@
+    ? [...checkedItems, ...childKeys]
+    : checkedItems.filter((x) => !childKeys.includes(x));
+
+export const mapTree = (item) => {
+  const checkedItems = getState().overview.checkedItems;
+
+  item.checkProps.checked = areAllDescendantsChecked(item, checkedItems);
+  if (
+    !item.checkProps.checked &&
+    areSomeDescendantsChecked(item, checkedItems)
+  ) {
+    item.checkProps.checked = null;
+  }
+
+  const retVal = { ...item };
+  if (item.children) {
+    retVal.children = item.children.map((child) => mapTree(child));
+  }
+  return retVal;
+};
```
What (I think) this code is doing is slightly odd:  for the current `item`, it decides whether it should be checked, unchecked, or "half-checked" -- based on the state of its children -- and then it proceeds to do the same recursively for each of the children...which means that it's going to visit each child once for each of its ancestors, which, if the tree is large, will be a lot of times!

What (I think) we should be doing is visiting the leaf nodes first, then determining from them how their parents should be set, and then returning to figure out how _their_ parents should be set, and so on.

Perhaps something like this would do it:
```javascript
export const mapTree = (item) => {
  const retVal = { ...item };
  if (item.children) {
    retVal.children = item.children.map((child) => mapTree(child));
    item.checkProps.checked = null;
    let seen = undefined;
    for (const c of item.children) {
      if (c.checkProps.checked == null) {
        return retVal;
      } else if (seen == undefined) {
        seen = c.checkProps.checked;
      } else if (seen !== c.checkProps.checked)
        return retVal;
      }
    }
    item.checkProps.checked = seen;
  } else {
    const checkedItems = getState().overview.checkedItems;
    item.checkProps.checked = isChecked(item, checkedItems);
  }

  return retVal;
};
```
That is, if the current `item` has children, evaluate each of the children (recursively).  Then, after setting `item`'s `checked` to `null`, look at the results of those evaluations:  if we find one which is `null`, then return immediately (with `item`'s set to `null`); otherwise, if we're looking at the first child (because `seen` is still `undefined`), remember whether it is set or not and move on to the next child; otherwise, if this child's `checked` is different from the last, then return immediately (with `item`'s set to `null`); otherwise, move on to the next child.  If we exit the loop without returning, then all the children are either checked or unchecked, and we set `item` to match.  Now, if `item` has no children, then we use `isChecked()` to set `checked`.  And, then we return.

So, we only call `mapTree()` once for each node, and we just look at the results on the way back up.  (We could replace my `for...of` loop with a `.every()` and a `.some()`, like you had, but that means going all the way through the children list twice, which is kind of unsatisfying -- with the `for...of` loop, we only check as many as we need to in order to get the right answer; and, this way, we don't need `areAllDescendantsChecked()` and `areSomeDescendantsChecked()` anymore.)

MVarshini replied:
That's really cool. Updated the code

webbnh replied:
Glad you like it!  😁

webbnh reviewed (approved):
OK, my principal objections have been addressed.

However, there are a couple of lingering things which I'd like to see you consider:
- My [suggestion to rename](https://github.com/distributed-system-analysis/pbench/pull/3464#discussion_r1290613754) a function was really more about a slight but significant change to the way the _function_ **worked**, which would remove a bit of code duplication...and, the name change doesn't really fit without the other code change.
- There are [two blobs of code](https://github.com/distributed-system-analysis/pbench/pull/3464#discussion_r1231546181) which are nearly identical, and I think it would be worth trying to remove the duplication.

Other than those, I've got a coding suggestion to address what looks like a nasty bit of inefficiency, and I've got a couple of nits, but there's no longer anything which I would block the merge for.

webbnh commented:
I don't see any changes since my last review.

MVarshini commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -458,35 +456,46 @@ export const parseKeySummaryforTree = (keySummary) => (dispatch, getState) => {
   for (const item in entity) {
-    const newKey = `${aggregateKey}${CONSTANTS.KEYS_JOIN_BY}${item}`;
-    if (!isServerInternal(newKey)) {
+    if (!isServerInternal(`${aggregateKey}${CONSTANTS.KEYS_JOIN_BY}${item}`)) {
+      const newKey = `${aggregateKey}${CONSTANTS.KEYS_JOIN_BY}${item}`;
```
The body of `for..in` should be wrapped in an if statement. This was triggering the lint complaint so reverted.

MVarshini commented:
@dbutenhof 
Problem 1: non-string values are not showing up
Stringified the values 

Problem 2: refresh problem
Updated the metadata

@webbnh 
Renamed the function back to `setChildNodes` . If I pass the parent which is an object I need to modify the function logic as it work on childNodes list

dbutenhof reviewed (approved):
Functionally, this looks good. I'm attaching sample output, after using Postman to set metadata as

```json
{
  "metadata": {
    "global.dave": "foobar",
    "global.foobar.dave": 12.50,
    "global.foobar.json": {"json": true, "value": {"integer": 123, "float": 1.4, "list": [1, 2]}}
  }
}
```

The odd thing about the `.toString()` function is that the output for the *list* type (`global.foobar.json.value.list`) looks more like a Python `",".join(data)` than a `str(data)`, which I'd expect to show that it's a list by including the braces. Still, it's readable, and I won't quibble about the format.

![image](https://github.com/distributed-system-analysis/pbench/assets/61761529/9772e6a4-afad-47c0-88de-a2a5659fa102)

webbnh commented on dashboard/src/actions/overviewActions.js:
```diff
@@ -523,3 +531,42 @@ const nonEssentialKeys = [
+ * @return {Object} - updated object with new value
+ */
+
+const setValueFromPath = (path, obj, value) => {
+  const [head, ...rest] = path.split(".");
+
+  return {
+    ...obj,
+    [head]: rest.length
+      ? setValueFromPath(rest.join("."), obj[head], value)
+      : value,
+  };
+};
```
This is very clever!  However, the fact that we split the path on entry and then rejoin it in the subroutine call is unsatisfying.  How about something like this:
```javascript
const setValueFromPath = (path, obj, value) => {
  const recurse = (plist, o, v) => {
    const [head, ...rest] = plist;
    return {...o, [head]: rest.length ? recurse(rest, o[head], v) : v};
  };

  return recurse(path.split("."), obj, value);
};
```
That is, we split the path once, on the initial call to the recursive function, and after that we just manipulate the array.

webbnh reviewed (approved):
> Renamed the function back to `setChildNodes`.

That's better than leaving the name mismatched with the body.

> If I pass the parent which is an object I need to modify the function logic as it work on `childNodes` list.

Right, but the modification is small.  The fact of the matter is, we need to mark all the nodes, and we do that by visiting each node and each of its children (if any); we can mark each node before (or after) we iterate over its children, or we can mark the nodes as we iterate over them.  The code as it stands marks the children as we iterate over them before recursing, which means that we need separate code to mark the parent.  If, instead, we mark the parent node in `setChildNodes()` (whatever we call the function), then we don't need separate code to mark it before the recursion begins.  The biggest difference is whether the recursive function receives just the list of children or the entire parent node (which contains the list of children) -- personally, this seems like a pretty small difference, but by passing in the parent node all the logic (the marking, the iteration, etc.) becomes encapsulated inside the function, and it doesn't have to be replicated.

In any case, the replication isn't very large, so we can live with it as is.  Everything else looks OK (although there are some other spots which could be DRYer...), so let's merge this!

(Although, I do have a couple of coding suggestions, for future reference.)

dbutenhof commented:
This is where our PR management policy is a bit loose. @MVarshini you're clear to merge the PR if you're happy; however I notice that you've "liked" several of Webb's latest comments and, in case you're planning to take action on those comments, I'm not arbitrarily merging for you. 😁

MVarshini commented:
> This is where our PR management policy is a bit loose. @MVarshini you're clear to merge the PR if you're happy; however I notice that you've "liked" several of Webb's latest comments and, in case you're planning to take action on those comments, I'm not arbitrarily merging for you. 😁

@dbutenhof  I noted those commands and will work on it in different PR. Please merge this PR :pray:
